### ISSUE TITLE
[Proposal]: Event hookup in object initializer

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/csharplang#5176]
# Event hookup in object initializer

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Allow event hookup like so in an object initializer:

```c#
var timer = new DispatcherTimer {
    Tick += delegate {},
    Interval = TimeSpan.FromSeconds(1d),
};
```

## Motivation
[motivation]: #motivation

It's not uncommon, especially in UI frameworks, to create objects that have both values assigned and which need event hooked up as part of initialization.  While object initializers addressed the first part with a nice shorthand syntax, the latter still requires additional statements to be made.  This makes it impossible to simply create these sorts of objects as a simple declaration expression, negating their use from things like expression-bodied members, switch expressions, as well as just making things more verbose for such a simple concept.

## Detailed design
[design]: #detailed-design

The existing https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#object-initializers will be updated to state:

```
member_initializer
    : initializer_target '=' initializer_value
+   : identifier '+=' expression
    ;
```

The spec language will be changed to:

> An object initializer consists of a sequence of member initializers, enclosed by { and } tokens and separated by commas. Each member_initializer designates a target for the initialization.  There are two forms of member_initializer, a [Simple assignment](https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#simple-assignment) form (`=`) and an [Event assignment](https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#event-assignment) form (`+=`).

> For the `Simple assignment` (`=`) form, the initializer_target identifier must name an accessible field or property of the object being initialized, whereas an initializer_target argument_list enclosed in square brackets must specify arguments for an accessible indexer on the object being initialized.  Each initializer_target is followed by an equals sign and either an expression, an object initializer or a collection initializer. It is not possible for expressions within the object initializer to refer to the newly created object it is initializing.  A member initializer that specifies an expression after the equals sign is processed in the same way as an assignment (Simple assignment) to the target.

> For the `Event assigment` (`+=`) form, the identifier must name an accessible event of the object being initialized.   A member initializer that specifies an expression after the `+=` sign is processed in the same way as an (Event assignment)[https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#event-assignment] to the target.  

> It is an error for an object initializer to include more than one member initializer for the same field or property.

## Design meetings

Note: there is no concern that `new X() { a += b }` has meaning today (for example, as a collection initializer).  That's because the spec mandates that a collection initializer's element_initializer is:

```
element_initializer
    : non_assignment_expression
    | '{' expression_list '}'
    ;
```

By requiring that all collection elements are `non_assignment_expression`, `a += b` is already disallowed as that is an assignment_expression.

**Comment by wrexbe** [comment]
I would use this feature. I want to know how it would handle property changes that trigger an event. 
What if something like this
```
var myText= new MyText{
    Text = "Hello"
}
myText.PropertyChanged += delegate {};
```
got refactored to 
```
var myText = new MyText{
    PropertyChanged += delegate {},
    Text = "Hello"
}
```
before the event would not trigger, and after it would get triggered.
Does the order matter? Would swapping them make the event not trigger?
```
var myText = new MyText{
    Text = "Hello",
    PropertyChanged += delegate {}
}
```

Should the language enforce the events get added last, or should auto refactoring tools try to add the events last, to avoid triggering them on accident?

**Comment by jnm2** [comment]
That's an existing question that applies to assignments of properties. I would be very confused if the syntax did not desugar to a series of statements in the order that the members are initialized in syntax.

**Comment by wrexbe** [comment]
Yeah that makes sense. Just worried about someone moving them around to be ABC order. I would probably keep them at the bottom most of the time.

**Comment by bartdesmet** [comment]
> Should we allow this for with { ... } expressions? I'm not sure.

While my scenario isn't about events, I recently encountered a case where the lack of compound assignment in `with` expressions sent me down an "expression-body-to-statement avalanche". The case was updating some deep record structure to increment a `TotalRuntime` accounting property of type `TimeSpan` with the `Elapsed` time obtained from some `Stopwatch`:

```csharp
var newStats = oldStats with { /* update a bunch of properties here ... */ Elapsed += sw.Elapsed, /* ... and more here */ };
```

**Comment by CyrusNajmabadi** [comment]
Thanks @bartdesmet I'll definitely bring up the question if we want to support this in `with` as well as supporting compound assignment for more than just events.

**Comment by jnm2** [comment]
> While my scenario isn't about events, I recently encountered a case where the lack of compound assignment in with expressions sent me down an "expression-body-to-statement avalanche".

This has happened to me as well.

**Comment by bernd5** [comment]
For me especially operator `??=` would be very useful in `with` expressions.

**Comment by jnm2** [comment]
`Click += new EventHandler(this.btn_Click) + new EventHandler(this.anyControl_Click)` would probably fall out automatically.

**Comment by najak3d** [comment]
Is there a good reason to not simply allow you to call ANY member method of the newly created object (and treat it the same as if you called it after construction)??   After all, property setters are really just calling the "Property_Set(value)" method; so why not simply extend this capability to allow calling ANY class method for the instance being constructed?

So for the event handler registration example, the answer is clear -- just have two entries:
var btn = new Button()
{
Clicked += EventHandler1,
Clicked += EventHandler2
}

It simply means that every statement inside the brackets ({ }) is treated as though it were prefixed by "btn.".

This is the Simplest solution/rule, and offers the maximum benefit.

**Comment by CyrusNajmabadi** [comment]
> Is there a good reason to not simply allow you to call ANY member method of the newly created object 

Yes. Methods imply state mutation as opposed to declarative construction. So I feel that it's much more natural for them to be after the instance read constructed. 

Also, that syntax is already allowed and is used for collection initialization. So it would need to be a syntax that would not be ambiguous with that.

**Comment by acaly** [comment]
> Methods imply state mutation as opposed to declarative construction.

This doesn't make sense to me. I agree that methods imply state mutation, but state mutaion is NOT opposed to declarative construction. Construction is really a kind of mutation in C#, because C# currently doesn't have a good mechanism to distinguish the two, until we have required init property and init-only methods. Also, although event registration should be seen as construction, a more general compound assignment is more like a state mutation. In any case, you shouldn't say something like "we don't want to add this because it is not initialization/construction".

**Comment by CyrusNajmabadi** [comment]
> Construction is really a kind of mutation in C#, because C# currently doesn't have a good mechanism to distinguish the two


I disagree. I think we do, and part of that is not blurring the lines more by using methods in these scenarios.

**Comment by najak3d** [comment]
The long-term trend has been that programmers want simpler (more terse) syntax.   Less typing accomplishes more (and thus less reading too).

For UI construction, some controls are written where the Children list is "Get-only" -- you can modify the contents of the list, but not "set the list" -- e.g. you can call AddChildren, or AddChild, but not "Children = new List<T>()".

So UI composition that SHOULD work without writing a full suite of hackish extension methods is the following:

new Grid()
{

>   Prop1 = propVal,
>   Prop2 = propVal2,
>   AddChildren(
>      new Button()
>     {
>        Clicked += EventHandler,
>        BindTo(binding),
>        ///... more 
>    },
>    new Button()
>    {
>       // compose Button2 here
>    }
>   )
> 

}

If a developer thinks "methods don't belong in initialization", then that programmer can decide to not-call-methods in the intializer.  But for the context where it is deemed grossly useful (e.g. UI composition) - allow it.  Why not?

**Comment by theunrepentantgeek** [comment]
> Why not?

**Comment by CyrusNajmabadi** [comment]
> The long-term trend has been that programmers want simpler (more terse) syntax.

This is not the long term trend at all.  And you can see very popular and very verbose languages that show that.

>  Less typing accomplishes more (and thus less reading too).

This is not a pro.  Clarity is what matters, not terseness.

> For UI construction, some controls are written where the Children list is "Get-only" -- you can modify the contents of the list, but not "set the list" -- e.g. you can call AddChildren, or AddChild, but not "Children = new List()".

This already works in C# today.  Just do this:

```c#
new Whatever
{
    Children = { ... }
}
```

This will add to the children list.  No need to `new` it up.

> If a developer thinks "methods don't belong in initialization", then that programmer can decide to not-call-methods in the intializer.

Or we can just not allow it at all if it's not going to be a good thing :)

--

note @najak3d please use normal github markdown markers around your code.  e.g. ```` ```c# ````.  I've edited your posts to use them properly :)

**Comment by CyrusNajmabadi** [comment]
> So UI composition that SHOULD work without writing a full suite of hackish extension methods is the following:

You can already do this in C# today with:

```c#
new Grid()
{
   Prop1 = propVal,
   Prop2 = propVal2,
   Children =
   {
      new Button()
      {
          Clicked += EventHandler, // will be supported by this proposal.
      },
      new Button()
      {
         // compose Button2 here
      }
   }  // end of Children
} 
```

**Comment by theunrepentantgeek** [comment]
> Less typing accomplishes more (and thus less reading too).

The two aren't strongly correlated. Shorter means more opportunity for confusion (dies the code you're reading so what you think it does). 

It's easy to go past concise to cryptic. 

Real world example:

What does SGTM mean?

Sounds Good To Me **or** Silently Giggling To Myself?

**Comment by najak3d** [comment]
I am very impressed with the community here.  This was my first day visiting these forums.  I'm a 51 year old programmer of C# since 2003, and want to see this language "rule all" -- I appears to me that C# has lost ground in the last 10 years, mainly due to Microsoft doing a piss-poor job of making Xamarin Forms competitive/good (they work worse than WPF).  Thus we have competitors like Flutter taking market share from what would have normally been done in C#.

One area where C# is hobbled is in the area of UI construction, which gave birth to XAML, which generally sucks.

It's unfortunate that C# has requires hackish full-suites of extension methods to achieve a notation that competes with Flutter.  I think this functionality shouldn't require this much work.

Yes, I agree that too-terse/overloaded statements is bad programming.  I tend to go the verbose route, not combining things into single lines (e.g.  method(anotherMethod(args)); IMO is normally bad IMO).

However, for UI composition, this domain/context really deserves better native support from the C# language itself.

It doesn't seem to be much of a stretch to simply say "any method can called from the Object Initializer block" -- and for safety, an order could be enforced such that "Init-Only Properties must always be called first", then after that, anything goes.

This would be: (a) Terse, (b) Clear, and (c) enable UI composition naturally.

It simply works similar to how VB "with" statements used to work, in that all operations are done on the main source object.  They of course all operate In-Order.    It should be the equivalent of making all  the same calls using a local variable.  In the Object Initialization block, it would just work like a "with" statement.  Very simple, concise, clear, and extremely useful for certain contexts.

For contexts where it's not appropriate, just don't use it.  But in short, even if you use it "inappropriately", what's the harm?  It works the same as if you just made all the same calls (in the same order) using a local variable assigned to the new object.   So it just amounts so a bit of very useful syntactic sugar, that makes UI composition syntax "natural/built-in", rather than an obtuse hack.

I think you are all awesome.  Thank you for your attention, and feedback.  This has been very encouraging to witness this energy , genius, and tone of the discussions.  Kudos to the whole group of you.

**Comment by acaly** [comment]
> I disagree. I think we do

Then what is the reason to allow general compound assignment, instead of just subscribing events, in object initializer? As I said, I agree that event subscription is important in object initializer, but I don't think allowing it for fields and property is anywhere better than calling methods, at least based on your initialization (or construction) vs mutation logic.

If I understand correctly,

```c#
var item = new Item()
{
  Value += 1,
};
```

does not distinguish whether the modification of `Value` is inialization or mutation. If you think you do, then the new syntax should only support init-only properties, not all set properties or fields. I know those are for back-compatibility. But that is exactly the problem here. Because of the back-compatibility, C# cannot have a clear difference between initialization and mutation from the callee side, or it will break old callers from modifying fields that is a mutation, not initialization.

Another reason I can think of why you say you do is that you think any modification of a field using compound assignment in object initializer can be seen as initialization instead of mutation. If that is the case, you are defining what is initialization/mutation by where the code is written, and that also means whatever we add to object initializer (e.g. method calls) will also be considered as initialization, which should be fine according to your initialization/mutation standard.

**Comment by najak3d** [comment]
It seems to me that "immutability" is currently enforced only by "readonly" modifiers or using "init" instead of "set" for properties.

Restriction: Readonly method - cannot use Init block -- all values must be set using a coded constructor.

Too Loose:  Init methods - currently Order-of-ops in Init Block do NOT enforce "init methods must be called first".  This seems unfortunate, because in a small way, this does not really enforce immutability, at least during the Init block.

As it currently stands though, the Init Block simply runs everything "in order" as you wrote it.  That's pretty simple.  So if we want to enforce "immutability" via Init Blocks, then we need to either:

1. Invent new notation (e.g. "InitFirst") to operate as Property Setter, which enforces order within Init blocks.
2. OR: Treat it as compiler error where an "Init only" setter is called AFTER anything except non-Init Properties.
(I'd vote for #2.)

So in this simple/sensible fashion, C#11 Init-Blocks can offer better "immutability" support, while also adding very useful syntax for nested composite construction (e.g. UI's).   Adds magnificent benefit, without confusion or downside.

**Comment by CyrusNajmabadi** [comment]
> It's unfortunate that C# has required hackish full-suites of extension methods to achieve 

> this domain/context really deserves better native support from the C# language itself.

You are being completely arbitrary. You reject one 'native' c# feature for being 'hackish', but then want some other features to do the same thing. 

Extension methods are native. They're part of the language. Avoiding features that have been around for 15 years and are widespread and fully embraced by the ecosystem just because you don't like them is not going to motivate is to create something new.

**Comment by najak3d** [comment]
> You are being completely arbitrary. You reject one 'native' c# feature for being 'hackish', but then want some other features to do the same thing.
> 
> Extension methods are native. They're part of the language. Avoiding features that have been around for 15 years and are widespread and fully embraced by the ecosystem just because you don't like them is not going to motivate is to create something new.

The NATURE of these extensions is hackish.  For example, Button already has a "Text" property, but because it does not "return Button", we cannot use it!  Therefore we have to create a NEW "Text()" method extension to use in it's place... to do the same thing, only it returns Button."

So it forces 100% replacement of all existing methods/properties, to support this syntax.  But if the Object-Init syntax were simply a bit more functional -- then we wouldn't need to use all of these hacked extension methods.

Extension methods to add functionality is a good idea.  And I'm glad they are available for scenarios like we are doing now -- because they allow us to "hack C# language" to essentially create a notation not currently supported by C# (but should be supported, IMO).

Extensions that are awesome are ones that extend collections to give you a "Count()" method for an IEnumerable, or the various other Linq extensions.  Those are all nice, but are ADDING functionality.

In our case, our methods are 90% NOT adding functionality, but are simply "redoing existing properties/methods" to make up for an inadequacy of the C# language itself (which is supported by Flutter/Dart and others).  That's why it's a hack, vs. the many other uses of Extensions, where it's not a hack.

IMO, in a way, though, most usages of Extensions are a bit hackish in nature.  In short, remove the "using ExtensionsNamespace" and your code suddenly breaks.   So it's a bit hackish/weird, and in many cases have been overused by some.  The best we can say about Extensions is that they are VERY helpful in overcoming awkwardness that would otherwise result without them, and so they are "good/useful" and we're glad they exist.

If you can write code that relies upon 1000 Extension methods, vs. writing code that needs almost ZERO extension methods, using nearly the same syntax -- it's preferred, by far, to have the code non-reliant upon these extensions.

**Comment by HaloFour** [comment]
@najak3d 

That's not a case of extension methods being hackish, that's a case of using extension methods to hack C# to smell like some other language because you want the style of that other language.  That is the hack, not extension methods.

There is no goal to make C# the language to "rule them all", especially not by blindly copying the syntax and idioms of other languages.  C# aims to be a good tool and to offer solutions to problems (note, not necessarily _the_ solution that you'd prefer).  If popularity and a user base comes with that, then great.  C# has a solid position on the TiOBE index and the 2022 rankings indicate an uptick in usage over the previous year.

**Comment by CyrusNajmabadi** [comment]
> The NATURE of these extensions is hackish. For example, Button already has a "Text" property, but because it does not "return Button", we cannot use it! 

I don't understand. Why don't you just set the Text property directly (either during construction or after?). It's already a mutable property. What do you need a method for?

**Comment by najak3d** [comment]
@HaloFour :
> That's not a case of extension methods being hackish, that's a case of using extension methods to hack C# to smell like some other language because you want the style of that other language. That is the hack, not extension methods.
> 
> There is no goal to make C# the language to "rule them all", especially not by blindly copying the syntax and idioms of other languages. C# aims to be a good tool and to offer solutions to problems (note, not necessarily _the_ solution that you'd prefer). If popularity and a user base comes with that, then great. C# has a solid position on the TiOBE index and the 2022 rankings indicate an uptick in usage over the previous year.

I'm glad we agree that the current methods used by MANY UI's is "hackish" (even Comet, I think is employing such a hack).   So to make C# capable of "composing UI's with simple notation", it requires this hack.   Extension methods ALLOW the hack.  I agree that Extensions themselves aren't innately "a hack", but are often used to create hacks, especially of this nature.

I do think .NET *should* rule-them-all.   The fact that it supports Extensions, has allowed C# to remain competitive against others like Flutter/Dart -- because this "hack" is mostly hidden from those using it.  Some 3rdparty generally "provides the extensions" and users don't have to do this for themselves.

For situations where C# is good, i.e. Client apps where you "write it once, and runs everywhere" - C# should rule.  But because Microsoft dropped-the-ball for over 10 years, it created a gap now being filled by Flutter/Dart.   Flutter/Dart is filling the gap and stealing market share in this domain.  MAUI is trying to steal it back, but appears mismanaged or understaffed, and has QA issues and is very late -- AND doesn't even provide a pixel-perfect UI solution (where UI operates/looks the same everywhere, as many apps prefer - i.e. one-user-manual).

So for client-apps, that need to run everywhere, C# should rule, without question (around 2006-2010 time frame, most new apps of this nature were written in C#; but now Xamarin-Forms has almost become a desert wasteland, in comparison).

Client-App development has been grossly trending towards MVU and UI's created in Code, not XAML.   C# requires a hack to do this, while many other languages do not.   IMO, this has become a clear C# deficiency, in the context where C# *should* rule.

**Comment by HaloFour** [comment]
@najak3d 

> I'm glad we agree that the current methods used by MANY UI's is "hackish" (even Comet, I think is employing such a hack).

No, I don't agree with this.  The hack is that the frameworks are intentionally avoiding the idiomatic approach for initializing types that has been provided in .NET and C# for 20+ years.

Simply put, stop trying to force the idioms of one language into another and you won't run into the friction.  That applies for any two languages or frameworks.

**Comment by HaloFour** [comment]
@najak3d 

> Flutter/Dart is filling the gap and stealing market share in this domain (e.g. 400,000 new Flutter apps since 2018)

Dart barely registers on the TiOBE index despite being around for over a decade.  COBOL and VBScript rank higher.  Even if language design was a popularity contest, which it is not, C# certainly has nothing to fear in this space.  C# is not going to turn into Dart to compete with Dart.

**Comment by CyrusNajmabadi** [comment]
@najak3d you still haven't provided any good examples of use cases these new language features are needed for. You've just claimed over and over again that they're needed. This is not an effective argument. 

I get that you think flutter is great and that MS needs to do more to complete there. But that's not an argument that means that the language needs to change. We could potentially do everything you've mentioned here in the language and not move the needle one iota. 

We need solid data (and real use cases) that demonstrate that it is the language side which is the problem here and that this is the right actually solution to that problem. 

Simply claiming it is by Fiat, and ignoring the counterarguments about how your existing use cases can already be solved is not helping your case. 

Please show real world cases that need this.

**Comment by najak3d** [comment]
@CyrusNajmabadi wrote:
> > The NATURE of these extensions is hackish. For example, Button already has a "Text" property, but because it does not "return Button", we cannot use it!
> 
> I don't understand. Why don't you just set the Text property directly (either during construction or after?). It's already a mutable property. What do you need a method for?

I've written this in other responses, and will reiterate it here.  There are several reasons that methods are needed for initialization:

1. To call Extension Methods created for simpler syntax (e.g. Pos = new Point(1,2)  vs   Pos(1,2) extension)
2. To call Extension Methods that do not exist on the target object (such as setting Grid Col/Row).
3. If API of class only provides a "SetMethod" rather than property setter (e.g. "SetPosition(point, reason)" vs. just "Position = point"

To name the main ones that I encounter.

Object-Init block does not allow for ANY extension methods to work (other than for Collection initialization maybe?).

And so the hacked solution now being employed widely is to create Extensions for EVERYTHING, required to make it work.

**Comment by CyrusNajmabadi** [comment]
>  Extension methods ALLOW the hack. I agree that Extensions themselves aren't innately "a hack", but are often used to create hacks, especially of this nature.

I'm going to be clear here *as a mod*. This argument is played out. You cannot simply claim this native approach to be a hack as a fact. You may not like it, but that's ultimately you just subjectively picking or choosing the existing, well established, language features that your are willing to use. 

C# is a pragmatic language. If there is a totally reasonable, very effective, existing solution, then we're not going to disregard that because you just are just adamant it's a hack. 

Such an argument is not convincing and does nothing to actually support your case. If you want this language change you have to show real use cases so we can establish 1. if those cases are important, 2. If they can be solved today, 3. If not, of your proposal is the right solution for them. 

Think about it this way. How would we even validate of our language change solves a problem if we don't have the problem in front of us to validate?  We could ship something and it could totally fail the moment you tried to use it for your domain.

**Comment by HaloFour** [comment]
@najak3d 

> I've written this in other responses, and will reiterate it here. There are several reasons that methods are needed for initialization:
> 
> 1. To call Extension Methods created for simpler syntax (e.g. Pos = new Point(1,2) vs Pos(1,2) extension)
> 2. To call Extension Methods that do not exist on the target object (such as setting Grid Col/Row).
> If API of class only provides a "SetMethod" rather than property setter (e.g. "SetPosition(point, reason)" vs. just "Position = point"
> 3. To name the main ones that I encounter.
> 
> Object-Init block does not allow for ANY extension methods to work (other than for Collection initialization maybe?).
> 
> And so the hacked solution now being employed widely is to create Extensions for EVERYTHING, required to make it work.

Of those three the only one I can kind of see is point 2, in that extension properties don't exist.

Beyond that the argument seems to be that you are working with frameworks that are going out of their way to work against the idioms and standards long established in .NET and C# and you are surprised/frustrated that by doing so that they have created unnecessary friction.  I would take those points to the designers of said framework as by following the proper conventions these issues almost entirely go away.  C# isn't going to shift its idioms because someone prefers to establish their own.

**Comment by najak3d** [comment]
@HaloFour :
> Dart barely registers on the TiOBE index despite being around for over a decade. COBOL and VBScript rank higher. Even if language design was a popularity contest, which it is not, C# certainly has nothing to fear in this space. C# is not going to turn into Dart to compete with Dart.

In a world where small apps are becoming more and more numerous, and devs look for the "right platform" to meet their needs, currently Flutter/Dart tops the list for that context.   
1. Write once, runs everywhere.  
2. UI is pixel-perfect -- test it once, and it runs the same everywhere.  (Xamarin Forms uses native controls, so it looks and runs differently everywhere).
3. Snappy/responsive.    (Xamarin-Forms is laggy.)
4. Allows Code-Based UI declaration, vs. XAML!    (Xamarin requires hack using 1000's of extensions for this)
5. Hot-Reload concept works faster and better.  (Xamarin Hot reload is bugged; often have to turn it off to prevent crashing)

So I'm speaking from the Xamarin perspective, having lived in this arena solely for the last 12 years.   Xamarin (and thus .NET/C#) has lost much market share to Flutter/Dart, and a few others.

I find myself hating the fact that "they've got it better than me" as I'm still trying to stick with Xamarin.Forms, soon MAUI.  C#/.NET used to RULE in this context; now they don't.

**Comment by HaloFour** [comment]
@najak3d 

So take that up with Xamarin and MAUI.  C# doesn't know nor care about those UI frameworks, or any UI frameworks.  What C# does provide is a standard syntax for object initialization and those idioms work just fine for UI frameworks as long as they follow the appropriate conventions.

**Comment by CyrusNajmabadi** [comment]
> Write once, runs everywhere.

This is not solved by a language change.

> UI is pixel-perfect -- test it once, and it runs the same everywhere. (Xamarin Forms uses native controls, so it looks and runs differently everywhere).

This is not solved bya  language change,

And this is true for all 5 of your points.

**Comment by najak3d** [comment]
@HaloFour :
> > I'm glad we agree that the current methods used by MANY UI's is "hackish" (even Comet, I think is employing such a hack).
> 
> No, I don't agree with this. The hack is that the frameworks are intentionally avoiding the idiomatic approach for initializing types that has been provided in .NET and C# for 20+ years.
> 
> Simply put, stop trying to force the idioms of one language into another and you won't run into the friction. That applies for any two languages or frameworks.

The "friction" doesn't bother me.  What bothers me most is that Flutter/Dart currently provides a "better platform for writing Client-apps that are written once, and run everywhere", and are stealing Xamarin market share in mass.  If Microsoft hadn't dropped-the-ball on Xamarin Forms -- this wouldn't be an issue.

Flutter/Dart are becoming more capable quickly, and are playing leapfrog with C# (similar to how C# played leapfrog with Java initially).    Similarly Java folks were mad when they lost so much market share to C# 20 years ago.   Now Flutter/Dart is doing the same thing now, to C#  (mostly because Xamarin.Forms suck, as does XAML).

Turns out... most people seem to prefer C# markup to XAML.  (similar to what Dart/others now allow)

**Comment by CyrusNajmabadi** [comment]
>  What bothers me most is that Flutter/Dart currently provides a "better platform for writing Client-apps that are written once, and run everywhere", and are stealing Xamarin market share in mass. 

@najak3d that may be true, but that doesn't motivate changes in the language just because. We'd need a strong argument as to why these changes to the language would move the needle in the slightest.

**Comment by CyrusNajmabadi** [comment]
> as I do not see the inherent dangers in allowing methods to be called from Object init blocks, which is the only barrier to enable the elimination of these hacked Extension methods.

We don't do things just because. We have literal finite resources for language change. So every language change has to provide immense value to be worth it. So this would have to solve actual real problems to warrant the time and effort involved to do all of this.

**Comment by najak3d** [comment]
@CyrusNajmabadi :
> And this is true for all 5 of your points. :)
> Your entire post appears to complaints about xamarin. None of your relates to the c# language afaict.

I am speaking from a Xamarin perspective - whose main focus has been on Client-Side apps.  Since Xamarin is "losing this battle", then in turn C# is losing that same battle.

Part of winning the battle is providing C# markup.  Many UI's have solved it via a HACK and extra code.   Problem is "solved with a HACK".

So it begs the question of "why does this require a hack?", especially when this hack would not be needed if we simply allowed method calls from inside the Object-Init block, which IMO, has no real downside, but rather only an upside.

**Comment by CyrusNajmabadi** [comment]
> To call Extension Methods created for simpler syntax (e.g. Pos = new Point(1,2) vs Pos(1,2) extension)

You don't need this. Just write:

`Position = { X = 1, Y = 2 }`

It's already terse, easy to read, idiomatic and well supported everywhere :-)

**Comment by HaloFour** [comment]
@najak3d 

> What bothers me most is that Flutter/Dart currently provides a "better platform for writing Client-apps that are written once, and run everywhere", and are stealing Xamarin market share in mass. If Microsoft hadn't dropped-the-ball on Xamarin Forms -- this wouldn't be an issue.

This has nothing to do with C# as a language.

> Turns out... most people seem to prefer C# markup to XAML. (similar to what Dart/others now allow)

I've yet to meet a UI framework in .NET that did not allow writing code instead of markup.

> Part of winning the battle is providing C# markup. Many UI's have solved it via a HACK and extra code. Problem is "solved with a HACK".

You not liking it does not make it a hack.  If you have to write extension methods to use a particular UI framework take that up with said UI framework.

**Comment by najak3d** [comment]
@CyrusNajmabadi :
> > What bothers me most is that Flutter/Dart currently provides a "better platform for writing Client-apps that are written once, and run everywhere", and are stealing Xamarin market share in mass.
> 
> @najak3d that may be true, but that doesn't motivate changes in the language just because. We'd need a strong argument as to why these changes to the language would move the needle in the slightest.

Your proposal is for "compound assignment"... here again, this can be done outside of the initializer block too.  So where is the problem?   What problem are you aiming to solve that isn't already being solved by less convenient notation?

I don't see a reason to "restrict what is called inside Object Init block".  The change I'm proposing is quite simple - it's mostly just to "remove existing restrictions" and make it more permissive.  Alleviate the restrictions, and make it more flexible.  The resulting code will not be confusing, as it runs the exact same as code would if you assigned a variable and then called the same methods directly after the object-Init block.    Treat Object Init Block mostly as a VB "With" block, with no (or few) exceptions.

**Comment by najak3d** [comment]
@HaloFour  - I thought you were the one who agreed that the 1000's of Extensions to effectively modify the C# language to make it support Dart-like notation was in fact "a hack".   Now it seems that you are saying "it's not a hack".   Which is it?   I think it's blatantly a hack, and I thought you agreed.

**Comment by HaloFour** [comment]
@najak3d 

> I don't see a reason to "restrict what is called inside Object Init block". The change I'm proposing is quite simple - it's mostly just to "remove existing restrictions" and make it more permissive. Alleviate the restrictions, and make it more flexible. The resulting code will not be confusing, as it runs the exact same as code would if you assigned a variable and then called the same methods directly after the object-Init block. Treat Object Init Block mostly as a VB "With" block, with no (or few) exceptions.

I disagree.  The notion of calling arbitrary methods seems entirely out of place for an object initialization block.  It's very much not intended to be like VBs `With` block (which has been roundly rejected for C#).

There needs to be significantly more reason to consider such a language change.

> I thought you were the one who agreed that the 1000's of Extensions to effectively modify the C# language to make it support Dart-like notation was in fact "a hack".

The desire to make a .NET framework model that of a Dart framework is the hack.  How you achieve that is irrelevant.  Stop trying to force the idioms of one framework into another and you won't experience this friction.

**Comment by najak3d** [comment]
Thank you for all of your attention and time.

@CyrusNajmabadi  - I see you are listed near the top on Github as a "Contributor" and so I imagine you have sufficient sway in the decisions made for the C# language.   I see about 120 Contributors in all on the list, and the vast majority of those Contributors who have seen my posts, do not seem to be supportive of my proposal.

So I think I've gotten my answer -- the answer appears to be a flat "no".   So I can stop my ranting and save us all some time.

I just find it a shame that C# has to be hacked to accommodate the preferences of perhaps a million client-app Devs, and as a result, watching C# lose out to rising alternatives such as Flutter/Dart.

Since Extensions have allowed us to hack C# in the desired manner, at least we are still able to accommodate, making my proposal "non-essential".    However, it should be noted that MANY of C# changes were also "non-essential", but were done mostly for purposes of simplification of notation, or other "non-essential" reasons.   So I would suppose that if enough people, like me, came to this forum to request this feature -- then it might gain momentum.

Currently I seem to stand mostly alone here, and thus there is no point to me continuing these rants. 

I thank the whole team here.   I love, love, love C#, and want to see it regain the market share lost to Flutter/Dart (and others).  That is a part of my overall mission here.

**Comment by CyrusNajmabadi** [comment]
> Your proposal is for "compound assignment"... here again, this can be done outside of the initializer block too. So where is the problem? 

Declarative initialization of events is not possible today. That's a core thing we aim to solve.  Similarly, declarative delta initialization from another object is not possible either.

**Comment by CyrusNajmabadi** [comment]
>  The resulting code will not be confusing, 

As I've already mentioned, this syntax is legal today and means something else :-)

So I do think that would be confusing.

**Comment by CyrusNajmabadi** [comment]
> So I think I've gotten my answer -- the answer appears to be a flat "no". So I can stop my ranting and save us all some time.

@najak3d this is not the case. I've spelled out explicitly what we would need *from you* to progress this argument further. 

Right now it is going in circles. Genuine use cases with real world apis need to be shown that demonstrate the strong need for this. 

Continuing to lament about flutter is not helpful. You haven't shown at all that this proposal of yours moves the needle in that discussion. 

Indeed, all your complaints seem to have nothing to do with the language and everything with the tooling.  So your argument comes across as nonsensical. In other words, it's a solution in search of a problem.

**Comment by CyrusNajmabadi** [comment]
> I thank the whole team here. I love, love, love C#, and want to see it regain the market share lost to Flutter/Dart (and others). That is a part of my overall mission here.

That's fine. But consider it from our perspective. Sometime says they want .net UI stacks to complete better against flutter.  Then they say something like: and thus GC is a hack, use ref counting instead. 

There's a huge disconnect between what is being proposed and the problem space they claim they want to see improved. 

That's the issue here. You've mentioned many issues with tooling (things like hot reload or pixel perfect), but then say we need arbitrary object invocation in initialization. 

There's a disconnect between the problems you are calling out and the solutions you are proposing

**Comment by najak3d** [comment]
My purpose here was mostly wanting to make sure this issue was heard -- and it was.   I don't think I have any other use cases that would help further my case.   Therefore I think the "answer is NO" is accurate.

I do bring up issues about the ecosystem in which I find myself, and want to make sure you understand that in the space where Xamarin-once-dominated, we are now quickly losing ground -- which means C# is also losing ground in this arena.

We are actively trying to regain that ground, and I wanted to make sure that one of the "necessary things we have to do, to compete" requires us to create a Hack, to alter C# syntax to mimic what the community wants.  So now you are aware (if not already), and I've made some noise.  I appear to be mostly alone here (or alone enough) to warrant dismissal of my request.

Maybe in the future, more will step forward, and this concept will gain momentum.  But I'm not counting on it.  It's really non-essential to my mission.  We have a way to move forward with C# "as is", and that is what we are doing.

**Comment by CyrusNajmabadi** [comment]
> to alter C# syntax to mimic what the community wants.

If the community wants this, the community needs to show the real use cases that this would really address. :-)

>  Therefore I think the "answer is NO" is accurate.

You are incorrect. That is not the answer, and it's wrong to keep stating that. 

The answer has been: this needs more if a defense and real answers as to why the existing options are not suitable and why this is the right 'solution' to an actual 'problem' that's been well explained. 

I've responded to many/all your points, but I've still been waiting on real answers to many questions on my part. Without that information I'm not seeing enough to demonstrate why this is important and should get focus over all the rest of the things we are working on. 

Thank you :-)

**Comment by najak3d** [comment]
I'll come back soon, with a few dumps of C# UI.Composition examples in Avalonia, so that you might witness the rightful "appeal" to this notation.  Not that it will change your mind, but may increase your awareness.

I believe the "init-block" (for easy stuff/setters) followed by the extension-block as was suggested by @TahirAhmadov should work for all cases I can produce, and would allow the reduction of Extension methods.   However, the dichotomy in blocks is not desirable to our target audience.   I believe they'll want the effect of being able to call any setter inside the same block, without differentiation.   Although, this approach is a decent "alternative" that I haven't seen done before, which can grossly reduce the Extension method count.

I'll be back sometime soon with more examples to post.

**Comment by jmarolf** [comment]
@najak3d I would love to see some specific examples of where the language falls down. I will admit the cases are not clear to me today.

Looking at darts construction: https://github.com/flutter/samples/blob/ae50cc192e1c0902d05dbd8e9a813ed0615efdea/desktop_photo_search/fluent_ui/lib/main.dart#L62-L96

I don't see how it that different (or at least vastly superior to) Xamarin's markdown syntax: https://github.com/xamarin/XamarinCommunityToolkit/blob/f892c5fb7a562dbeb42bb724abc2ebaa780eca5e/samples/XCT.Sample/Pages/Markup/SearchPage.cs#L40-L74

Or the upcoming comet syntax for MAUI: https://github.com/dotnet/Comet/blob/87b0d575d62215b87725875a2f51ad6b6f9a4cbe/sample/Comet.Samples/Views/BasicTestView.cs#L42-L56

But I have not written a production app in all three frameworks so there are likely things I am missing. It may be best to have this as a separate discussion on https://github.com/dotnet/csharplang/discussions to start. if @Clancey or any folks from @dotnet/maui-internal-contributors want to chime in and talk about specific areas where the language has made things harder that would be extremely helpful.

As it stands today this discussion seems a bit orthogonal to @CyrusNajmabadi's proposal and I would prefer that we discuss potential problems of compound assignment in object initializers.

**Comment by najak3d** [comment]
The main "fall downs" is that if you need to do anything other than setting simple properties/fields, you cannot do it without writing EVERY possible thing you might want to do as an Extension Method to enable the very popular Fluent-Pattern.

Since Fluent-Pattern is so popular and accepted, IMO, it begs the question "why does it require so much extra code to enable this syntax", when it could be fully enabled if C# simply enabled you to call methods on the newly constructed class via ".{methodName}(...)" notation.

Xaml Markup, Comet, WPF -- all require considerable amount of "Extension methods" to achieve this very popular/widespread Fluent syntax.     The Fluent-Pattern syntax itself is quite non-intuitive, and works like a hack (non-standard to how you generally write good C#), forcing you to write methods that non-intuitively *must* return the Object itself, so that the next command can then operate on it.

While the ".{methodName}(...)" notation requires ZERO odd-ball methods/extensions to achieve an intuitive, simple, and popular behavior, useful for initializing and instantiating your full objects inside of a composite/nested format.   It still seems obvious to me that C# should allow this natively, without hack-ish extensions.

For example, to apply a "style to UI elements" you can call a "With(Action<UIElement> stylingAction)" on a newly created UIElement, and this method will apply Font Size/Style/Color, and Margins/padding/etc, and so requires a method call.   And for other objects you might call "Enable()" or "initialize()" right after constructions... which is also a method.   And one of the biggest reasons to call methods is that WPF, Xaml, Avalonia, etc -- make heavy use of "AttachedProperties" which aren't native C# Properties, but instead are Read/written via Method calls -- so currently you can't set ANY AttachedProperties from inside a standard C# Init block.

And so, in order to achieve the most attractive syntax for Fluent-Pattern inside C#, requires an EVERYTHING approach, which means that even to set individual properties, you have to create a Fluent-style Method for EACH simple property as well, such "SetHeight(height)" which returns back the object. 

But we note that the proposal on new Roles/Extensions, however, helps with this situation very much -- and so we're looking forward to that one being implemented, as it'll provide a better solution than the current solutions employed to enable Fluent Pattern.

**Comment by CyrusNajmabadi** [comment]
> it begs the question "why does it require so much extra code to enable this syntax"

It doesn't require "so much extra code". Fluent can be done with just a dot and the method name. That is an extremely small amount of extra code.  

The primary issue seemed to stem from the subjective distaste you have for that syntax here. 

Note, your own syntactic proposals here either need the same amount of "extra code", or *more*. Your code would still need a dot and the method name, *and* you'd have to follow the call with a comma as well.

**Comment by CyrusNajmabadi** [comment]
>  It still seems obvious to me that C# should allow this natively, without hack-ish extensions.

Describing one is the most common and will understood language constructs that have been around more than 15 years as "hackish" is not an argument. You have subjectively chosen to not use the feature that was added where this sort of use case was one of the scenarios it was designed for. 

Stop calling it hackish when all you really mean is "I don't like it, and I want something equivalent that I just personally think it's nicer".

**Comment by najak3d** [comment]
I cannot stop calling it "hackish" in the manner of what you have to do to "enable Fluent Pattern"..    Once you write the abundant extensions to cover ALL things you want to set (including simple properties) - now you can write simple Fluent syntax.   Fluent Pattern was a "hack" of the C# language to make it do something that you (and many others here) seem dead set against allowing C# to do natively without the hack.  And since C# doesn't allow this natively, we all continue to do this hack, so that we can enjoy the syntax that so many have grown accustomed to.

I'll probably never ever ever see the harm, confusion, nor danger of enabling this type syntax:
(because there truly is not confusion, danger, or harm here)

```c#
new Label()
{
Text = "This should be legal",   // simple property
Font = Fonts.Tahoma,  
.SetStyle(TextStyle, IsItalic),   // Init method, but have more than one parameter
.SetRowCol(2, 3),     // Sets a virtual Attached Property
.Bind(ModelProperty, Mode.TwoWay)   // multiple arguments
}
```
How could this code here ever be unclear or confusing, or ambiguous?   It's ideal, and enables the Fluent-style initialization without ANY added extensions!  (all of the methods called above are just methods that *already* exist)

**Comment by CyrusNajmabadi** [comment]
At this point @najak3d were at an impasse. I ask that you make another discussion as this is entirely off topic for how proposal and you are not addressing counter points and questions. Consistently coming back to a position that you state is unwilling to consider other prescribed perspectives means there's nothing more to discuss on this.

**Comment by najak3d** [comment]
Alright, let's just talk about your proposal here, to allow this notation:
```c#
var timer = new DispatcherTimer {
    Interval = TimeSpan.FromSeconds(1d),
    Tick += (_, _) => { /*actual work*/ },
};
```

This is ALREADY the equivalent of in C#, using Fluent-Pattern.  The steps to do it are widespread and popular -- it's called Fluent-Pattern.  All you have to do is write 100's of Extension methods so that you can do what you want for initialization logic, including adding an extension that permits you to add event handlers.

The resulting Fluent-Pattern style logic simply looks like this:

```c#
var timer = new DispatcherTimer()
    .Interval(TimeSpan.FromSeconds(1d))
    .Tick((_, _) => { /*actual work*/ })
```

So there is ZERO need for your proposal.  It's a useless proposal, that stems from YOU NOT LIKING FLUENT PATTERN.  Why is it that you don't want to simply use Fluent Pattern for this?  It's a wonderful thing, you say.

By your own arguments, you've FULLY de-legitimized your own proposal here.

**Comment by najak3d** [comment]
I just want to take this "wonderful Fluent Syntax" and wrap it in "{ }" so that you can precede your Fluent style method calls with simple property setters, AND requiring ZERO extension methods to make it work.

Fluent is such a wonderful/popular pattern, so why does C# require a hack to make it work? (a hack that even YOU don't want to do for your own issue)

**Comment by theunrepentantgeek** [comment]
Assume for the sake of argument that C# was modified so that all methods could be called within initializer syntax. 

What effect would this have on the overwhelming majority of types that are _not designed with a fluent interface_?

In a few minutes of thought I came up with several nasty scenarious - not just code that's _difficult_ to read, but code that _doesn't do_ what it looks like it does, and code that would fail at _runtime_.

Language design must address more than "which useful scenarios does this enable" - it also needs to cover "which nasty outcomes does this enable" too.

**Comment by CyrusNajmabadi** [comment]
>  Why is it that you don't want to simply use Fluent Pattern for this?

I never said I didn't. Stop strawmanning.  I was explicit that I thought that absolutely should be part of the discussion.

**Comment by najak3d** [comment]
@theunrepentantgeek  - C# enables plenty of nasty looking code, if a bad developer decides to create a cryptic mess.  Unsafe code can look nasty - yet we allow it.  Extensions can literally make ANYTHING possible - yet we allow it. 

I would only be concerned if something enabled by C# introduced a "new confusion" where you could think you are doing one thing, but then something else happens, that is hard to figure out.    It's one reason C# broke the awful mistakes of C++, such as with memory leaks, and also with statements like:  "If (a = b) {}" ... where the coder meant to use "==".

C# still allows unsafe code, which opens doors for memory leaks, but we allow it.

So I'm interested to see your "reasonable/common examples" that you perceive to be the risk for simply allowing the ".method()" notation to be called on a newly constructed object, behaving the exact same as if you had done long-hand coding of the same thing.  (where you assign a local variable, then call all those methods using the local variable reference for each call)

I don't see how you would ordinarily do anything "confusing" producing unexpected results for ordinary looking code.   Please provide your best 3 examples here, so that we can know what you are talking about.

**Comment by najak3d** [comment]
Also, what do you mean by types "not designed with a fluent interface"??   Every class has a "fluent interface" if you simply write the extensions.   No good class SHOULD EVER have a native Fluent-Interface, because the fluent-pattern-methods are nonsensical in nature -- thus, as is, since it's a hack -- fluent-interface methods should ALWAYS be added via Extensions, or by creating a dedicated "Builder class" that provides the fluent interface for the composition context.

OR -- C# could simply allow it, because it's very sensible and non-confusing/non-dangerous.

**Comment by CyrusNajmabadi** [comment]
> because the fluent-pattern-methods are nonsensical in nature

this position is going to go nowhere.  You are effectively stating an unwillingness ot accept or use a bog-standard pattern that .net and other APIs have had for 15+ years.  Any argument that comes down to your dislike of this aspect of our ecosystem isn't helping as it effectively is you stating htat we need to be constrained by your own personal foibles.

**Comment by CyrusNajmabadi** [comment]
@najak3d I left you some info on Discord :)

**Comment by najak3d** [comment]
> > Why is it that you don't want to simply use Fluent Pattern for this?
> 
> I never said I didn't. Stop strawmanning. I was explicit that I thought that absolutely should be part of the discussion.

I'm not strawmanning here.   You were blatant in your response to me that my proposal has no ground to stand on, because it could be resolved using 100's of Fluent-extensions, to enable the simple syntax.   You've cast me as someone who is petty and only proposing a change because I don't like the "current widely adopted/understood Fluent syntax".   Yet, you are EXACTLY THE SAME with regard to your own proposal.  Your proposal could be resolved by Fluent Pattern - so why raise this issue at all?

So you've asked me to "defend my position" -- why do I need to?  You ALREADY NEED TO DEFEND MY POSITION, to promote this proposal here.

I KNOW why Fluent would suck as a solution for yours -- because it's an awkward Hack -- which is only used because it's better than not using Fluent method.  And so we've adopted and standardized this Hack -- just because it's widely adopted, doesn't mean it's any less of a hack.

I'd like C# to support Fluent style initialization in a natural easy Hackless fashion, as it should.  There's just not a strong defense for NOT simply enabling the ".method()" notation inside an init block.... at least not that I've ever seen.

**Comment by najak3d** [comment]
> > because the fluent-pattern-methods are nonsensical in nature
> 
> this position is going to go nowhere. You are effectively stating an unwillingness ot accept or use a bog-standard pattern that .net and other APIs have had for 15+ years. Any argument that comes down to your dislike of this aspect of our ecosystem isn't helping as it effectively is you stating htat we need to be constrained by your own personal foibles.

YOU ARE DOING THE SAME THING.  That's my point.  YOU ARE UNWILLING TO ACCEPT FLUENT AS A SOLUTION TO YOUR OWN PROPOSAL.   Am I wrong here?    You and I are in the same-boat, but you are pretending that I'm being "petty" while you are not....  but we're in the same boat here.

**Comment by CyrusNajmabadi** [comment]
>  Yet, you are EXACTLY THE SAME with regard to your own proposal. Your proposal could be resolved by Fluent Pattern - so why raise this issue at all?

So that can be discussed.  It will be part of the convo about why we would need this if it's already possible.  Indeed, tha'ts part of my discussion.  I do *not* view extensions as a hack, so they will be fairly compared against the alternatives and the proposal will hve to stand against that.  I do not dismiss existing language options as being unsuitable as 'hacks'.

**Comment by najak3d** [comment]
You made this proposal because you think "It's likely a good idea".  How did you ever come to this conclusion, if you so firmly believe that "Fluent is awesome" and so "let's just write 100's of extensions for each scenario where we want to enable event handling inside the init block"????    Why on earth did you even make your "absurd proposal" given your firm beliefs that Fluent is so awesome/accepted/appropriate?

The reason you don't want to use Fluent is that Fluent is annoying to implement, because it's so of "all or nothing".   You can't just write ONE METHOD to enable event handling, and be done.  You pretty much have to wrap ALL of the possible properties/methods, else your fluent-options are incomplete.

**Comment by CyrusNajmabadi** [comment]
> You made this proposal because

I made this proposal to bring as a potential option to the LDM to see if the feature is worthwhile and carries enough value for us to support.  

> You made this proposal because you think "It's likely a good idea". How did you ever come to this conclusion, if you so firmly believe that "Fluent is awesome"and so "let's just write 100's of extensions for each scenario where we want to enable event handling inside the init block"

I'm not going to engage if you're not acting in good faith.  You quoted three things there that i can find no record of ever saying myself.  This is a strawman.  You're expecting me to defend myself against things *i didn't say*.  

> The reason you don't want to use Fluent is that Fluent is annoying to implement, because it's essentially "all or nothing". You can't just write ONE METHOD to enable event handling, and be done. You pretty much have to wrap ALL of the possible properties/methods, else your fluent-options are incomplete.

this is not true.  I don't believe i have to wrap all the possible properties/methods.  For example, i can just pass hte instance along and do whatever i want with it.  One issue though is that this is not compatible with `init` properties.

**Comment by najak3d** [comment]
Suggesting Fluent as a viable alternative to resolve your issue, is a silly notion, and is a practice that shouldn't be made "more widespread".

I picture the fluent hack to be like this:  Imagine a world where you aren't allowed to fly with ANY medications at all...  No medications allowed on a plane -- either carry on or luggage!.... so if you want to bring medicine with you, you must drive, and it takes 10x longer, and more cost/risk.   But it's YOUR ONLY OPTION, other than not going.   So MILLIONS of people need medication and so choose to DRIVE, because it's their only option.  But driving is better than NOT GOING... so they drive.

Then along comes someone, named kajan4D, who says "can't we just allow meds to be package into the luggage along with cosmetics"?   And the airline says "why are you rejecting the widely accepted practice of driving???   It's widespread, and understood for 15 years, so it must be good; in fact, let's consider using this driving solution for MORE things."

In my view, that's what is happening here now.   I'm being told to "keep driving because it works, and is widely practiced, and therefore there is nothing wrong with it".

Nobody should seriously consider Fluent as an acceptable hack to enabling Event handling.   Nor should it be used for calling a plethora of other initialization methods.

**Comment by TahirAhmadov** [comment]
I will repeat my earlier suggestion again. Given that methods can't really be called before all properties are set, because of `init`/`required` scenarios, even if they were allowed to be called, they would have to be called at the end of the init-block. And instead of adding method invocations to init-blocks, why not just do this:
```cs
// one "fluent" extension method to catch all
public static T Configure(this T This, Action<T> action) { action(This); return This; }
var btn = new Button
{
  Text = "asd",
  Click += this.btn_Click,
}
  .Configure(b =>
  {
    b.WithStyle(style); // call existing methods which return void or whatever
    b.Enable();
    // etc.
  });
```

**Comment by CyrusNajmabadi** [comment]
> Suggesting Fluent as a viable alternative to resolve your issue, is a silly notion,

It is not, and i really need to make it clear that you need to check your perspective on this stuff.  It is absolutely appropriate and reasonable for people to suggest alternatives and to consider if "the status quo" is just more preferable than doing anything here.

We've survived without this for 20 years.  We will absolutely be considering if we can keep on going without anything else here.  And we will consider if hte existing solutions are viable.  

I strongly suggest we chat.  You keep presuming that we should view this space just like you do.  And that sort of attitude is not going to go anywhere in terms of gaining consensus.  I can tell you from 20 years doing this, it is not a good idea to tell people that their position is silly, as opposed to spending time and effort to understand how htey think about things and to work with them to find solutions that satisfy their concerns as well.

**Comment by najak3d** [comment]
It's unreasonable to believe that you were inspired to write a proposal to change C# that you didn't already think was "likely a good idea", despite seeing "how awesome Fluent already is".   Why even bother with this proposal if you think Fluent is so great.

I get that you are open-minded and not drawing your final conclusions yet.  But -- the fact remains is that you already knew about Fluent as a solution -- but STILL made this proposal, because you clearly perceived that your proposal was "Likely better" than using Fluent.

Yet when it comes to my proposal, you seem to instead champion Fluent, and mildly insult me implying that "Fluent just doesn't match my taste", without realizing that I don't like the "100's of hackish extensions required to enable the syntax."  It's 100's of hackish methods that shouldn't ever NEED to exist in the first place, if C# simply allowed this very clear/sensible/non-dangerous syntax in the first place.

**Comment by CyrusNajmabadi** [comment]
> Nobody should seriously consider

@najak3d this is not acceptable.  Stop telling people how they should assess as situation.  At the very least recognize that there is no viable path forward that involves browbeating people to assess things like you do.  Please come to discord so we can chat about this.

**Comment by CyrusNajmabadi** [comment]
> Why even bother with this proposal if you think Fluent is so great.

To have the situation be discussed and evaluated.  Because i strongly want to get the views of the rest of hte ecosystem, as well as the rest of the LDM to see how they feel about this.

I have come with many proposals over the years that i've had varying levels of supportive thoughts around.  I've even brought things that i did *not* think was a good idea, just so it could be assessed by others to get their views.  Please stop presuming that this process works how you want it work and that others need to feel the same way about these topics as you do.

The reason we have this process and the reason why we have so much discussion and communication here is because there are so many different perspectives and so many different views on what is or isn't valuable for the language.

**Comment by CyrusNajmabadi** [comment]
> without realizing that I don't like the "100's of hackish extensions required to enable the syntax."

That was a decision on your part.  No one here *ever* suggested or implied you should do that.

**Comment by najak3d** [comment]
> I will repeat my earlier suggestion again. Given that methods can't really be called before all properties are set, because of `init`/`required` scenarios, even if they were allowed to be called, they would have to be called at the end of the init-block. And instead of adding method invocations to init-blocks, why not just do this:
> 
> ```cs
> // one "fluent" extension method to catch all
> public static T Configure(this T This, Action<T> action) { action(This); return This; }
> var btn = new Button
> {
>   Text = "asd",
>   Click += this.btn_Click,
> }
>   .Configure(b =>
>   {
>     b.WithStyle(style); // call existing methods which return void or whatever
>     b.Enable();
>     // etc.
>   });
> ```

Why not simplify that syntax to simply this?
```c#
> var btn = new Button
> {
>   Text = "asd",
>   Click += this.btn_Click,
>    .WithStyle(style), // call existing methods which return void or whatever
>    .Enable()
>   };
```

It's considerably less typing and less tedious -- without introducing ANYTHING confusing or ambiguous?

**Comment by CyrusNajmabadi** [comment]
>  without introducing ANYTHING confusing or ambiguous?

Nothing is confusing or ambiguous about that solution for me.  It's literally some of hte most vanilla C# that has been supported and clear for 17 years now.

**Comment by najak3d** [comment]
> > without realizing that I don't like the "100's of hackish extensions required to enable the syntax."
> 
> That was a decision on your part. No one here _ever_ suggested or implied you should do that.

It just seems like you are wishy-washy on your stance.   Wasn't it you that said "Fluent-Pattern is widespread for 15 years"... I can either choose to have nice Fluent-Syntax for our users, or not.   So we choose to suffer-the-pain, to give our users a better experience.   But there's just no good reason that "Fluent-Pattern should require this added Pain, just to enable it" -- not when there is a very simple/viable/sensible solution that would make this pain go away 100%, without introducing any new viable dangers (that I have ever seen).

**Comment by TahirAhmadov** [comment]
Really though, I'm not sold on the idea that one needs to call all that many methods to construct UI controls. I've been doing WinForms and ASP.NET WebForms for years and WPF for a year or two, and I never had the need to call methods on the controls to initialize them. Even when I created custom WinForms controls, I added properties for which the designer auto-generated init code without issues, using the appropriate attributes and converters and such. Same thing with ASP.NET ASPX markup language and WPF XAML - both of those markup languages work fine because all you need 99.99% of the time is to set properties.

> Why not simplify that syntax to simply this?

I guess the question is, is it worth the effort to add this, if all you are saving is literally 3 lines - the `.Configure(b=>`, `{` and `})`. I understand and agree that 3 lines can get multiplied by many times, and that goes back to what I wrote above - I would like to see concrete examples of things that need to be methods. You brought up a couple more:

> For example, to apply a "style to UI elements" you can call a "With(Action stylingAction)" on a newly created UIElement, and this method will apply Font Size/Style/Color, and Margins/padding/etc, and so requires a method call. And for other objects you might call "Enable()" or "Initialize()" right after constructions... which are also methods (not properties). And one of the biggest reasons to call methods is that WPF, Xaml, Avalonia, etc -- make heavy use of "AttachedProperties" which aren't native C# Properties, but instead are Read/written via Method calls -- so currently you can't set ANY AttachedProperties from inside a standard C# Init block.

Let's go through these one by one. The `With(Action stylingAction)` is an interesting approach, however, if it was me, I would have a strongly typed interface, `IStyleAdder` or something, and then the control can have a property of that type; but you can even create a property of type `Action` if you like. `Enable()` is begging to be changed to `bool Enabled { get; set; }`. I don't understand what `Initialize()` means - what is it supposed to do? Regarding attached properties, those are a bit more finicky; personally, I dislike that whole approach - after all these years it's still weird to me; however, my opinion of attached properties aside, if you need to set those, check out the discussion Fred (I think) started about adding attached properties as part of the roles feature.

**Comment by CyrusNajmabadi** [comment]
>  So we choose to suffer-the-pain

I would suggest coming to discord.  On a personal level, i will tell you that at this point is appears to me that you're not interested in even trying to reach any sort of consensus or to get someone like me to champion your position.  That you can only categorize our views as causing people ot "suffer the pain", or insist on strawmen like claiming that we're telling people to write hundreds to thousands of extension, or stating that i'm not willing to consider alternatives like fluent myself (where i am), means that you're not really interesting in engaging.  

I have been trying to explain my positions and explain the concerns i have with yours.  It does not feel like you are willing to accept that other people may feel differently than you on these topics, and that means there is very likely no paths forward here.

If you want to come to discord, we can discuss further how best to potentially get interest (and a champion) for your proposal.

**Comment by najak3d** [comment]
@CyrusNajmabadi  wrote: "Yes. Methods imply state mutation as opposed to declarative construction. So I feel that it's much more natural for them to be after the instance read constructed."

Which makes for a simple rule that would cause a compiler error -- it treats the init block the same as it does now, UNLESS it comes across a ".method(..)" -- at which point the object becomes "fully constructed" and the next lines of code become the exact equivalent to long-hand calling of methods AFTER the init-block.   Putting inside the init-block is just for short-hand... because it's still 100% clear what you are trying to do. 

AND it enables us to support Fluent-Pattern without any hacks.  It would then "just work".

**Comment by CyrusNajmabadi** [comment]
> Putting inside the init-block is just for short-hand...

I get that.  I'm saying, i find extremely marginal benefit for such a shorthand.  I'm not a believer in the idea that this is common/widespread.  Furthermore, your claims that this causes thousands of extensions to be written just seems off to me as it hasn't been explained why you'd need more than a *single* extension here

```c#
public static T Init<T>(this T value, Action<T> init)
{
    init(value);
    return value;
}
```

Why do you need thousands?

**Comment by najak3d** [comment]
@CyrusNajmabadi wrote: "From my perspective, it works fine today."

Which seems to also mean, "from YOUR perspective, adding event handlers via Fluent Pattern ALSO works fine today"-- because it does -- so long as you don't mind converting everything over to Fluent pattern, or tacking them on AFTER the init block.  Both of those "work fine today" as well.

I've never heard of someone saying "I can't add event Handlers after construction" -- there's no issue here.  They obviously can do this already in a variety of ways.   "I works fine today."  ...  So why did you make a proposal to fix something that already "works fine today"????

Fluent does work fine; but it's still extra work, and a hack.  We'll continue to use it, unless C# ever matures to a point where we don't need this hack to have this simpler syntax.

The new Roles/Extensions does offer a solution that is superior to current Extensions, but still isn't as good as simply allowing this syntax to exist inside the init block.

**Comment by HaloFour** [comment]
@najak3d 

This is the kind of friction that you get when you take one API and try to force it to follow the conventions of a completely different API.  Stop doing that and those "hacks" disappear.  If you want MAUI or .NET UI APIs to follow a more fluent API design, maybe take that up with them.  This is not a language concern, and aside from event handlers in initializers I've yet to see anything in Flutter that doesn't fit in the existing design of C#.

**Comment by jmarolf** [comment]
wew, guess I'll wade in here again

Object construction is one of the most common things done in C# today (if not _the_ most common) so improving it would certainly be worthwhile. However, I would like us all to be very specific about the problems we see today with object construction before jumping to any conclusions about what the design to fix them should be.

From my perspective we have two kinds of objects that are commonly constructed today.

- plain old data (POD)
- "rich" objects

For plain-old-data the records feature was added with the hope that it would make their construction simpler. I would say this has mostly been a sucess though there are areas for improvement. Most importantly to me POD is very often serialized and in the world we live in today that often means setting properties.

In the world of POD that means constructors are not very useful so if I declare a `record` like this:
```csharp
public record WeatherForecast(DateTimeOffset Date, int TemperatureCelsius, string? Summary = null);
```

It's not likely that is constructor ever gets called in real code because it most likely always going to be initialized like this:

```csharp
WeatherForecast? weatherForecast = JsonSerializer.Deserialize<WeatherForecast>(jsonString);
```

which leads us to the place we have today where most POD type are written like this:
```csharp
public record WeatherForecast
{
    public DateTimeOffset Date { get; init; }
    public int TemperatureCelsius { get; init; }
    public string? Summary { get; init; }
}
```

The fact that PODs need to be serializable and often represented in the system as json, xml, etc leads to proposals like [Dictionary Literals](https://github.com/dotnet/csharplang/issues/414) and of course the extremely exciting [Roles and extensions](https://github.com/dotnet/csharplang/issues/5497).

However, those same requirements also mean that fluent construction of these types must not be required because asking a serialization framework to call your fluent apis is too much complexity to put on them imho.

```csharp
var forecast1 = new WeatherForecast{
    Date = DateTimeOffset.UtcNow,
    TemperatureCelsius = 10,
	Summary = "Redmond Forecast"
};

var forecast2 = forecast1 with {
    TemperatureCelsius = 15,
	Summary = "Kirkland Forecast"
};

var forecast3 = forecast2
	.WithTemperatureCelsius(12) // how is a serialization framework going to know to call these?
	.WithSummary("Bellevue Forecast");
```

My general opinion is that records and with-expressions are supposed to primarily apply to POD-like scenarios and so I am not concerned about fluent api syntax there. My expectation is that [Required Properties](https://github.com/dotnet/csharplang/issues/3630) and other proposals that focus on improving simple object construction will help more people in for now.

All my concerns about POD and fluent apis don't apply to "rich" objects though (note I am only using quotes here because rich can mean so many things). UI is a good example of rich objects imho. They encapsulate state and need to manage much more than plain-old-data. For construction of these types of objects we saw the fluent API pattern develop in C# 17 years ago.

But we don't just use fluent apis in UI frameworks. We also use a fluent builder pattern _today_ in the default ASP.NET template:
```csharp
var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddRazorPages();

var app = builder.Build();

// Configure the HTTP request pipeline.
if (!app.Environment.IsDevelopment()){
    app.UseExceptionHandler("/Error");
}
else{
	app.UseDeveloperExceptionPage();
}

app.UseStaticFiles();

app.UseRouting();

app.UseAuthorization();

app.MapRazorPages();

app.Run();
```

but I am not totally convinced that allowing method calls here would make things look much cleaner or be easier for developers to use and understand:

```
var app = WebApplication.CreateBuilder(args){
        .Services.AddRazorPages()
    }.Build(){
        .Environment.IsDevelopment()
            ? .UseDeveloperExceptionPage()
            : .UseExceptionHandler("/Error"),
        .UseStaticFiles(),
        .UseRouting(),
        .UseAuthorization()
        .MapRazorPages(),
    };

app.Run();
```

or if we allowed "withers" to work here
```csharp
var builder = builder with { .Services.AddRazorPages() };

var app = builder.Build();

// Configure the HTTP request pipeline.
if (!app.Environment.IsDevelopment()){
    app = app with { .UseExceptionHandler("/Error") };
}
else{
	app = app with { .UseDeveloperExceptionPage() };
}

app = app with {
    .UseStaticFiles(),
    .UseRouting(),
    .UseAuthorization(),
    .MapRazorPages()
};

app.Run();
```

To be clear this may not be the best example, but I think it's relevant. Whatever we do should make sense for all the various domains people use the language.

It sounds like what you want @najak3d (though it's always dangerous to assume here...) is a [pipe-forward operator](https://github.com/dotnet/csharplang/discussions/74).

Here is an example in F# is pipe-forwards for their web template (though this is not apples to apples as the F# template has more features)

```fsharp
let todosApi =
    { getTodos = fun () -> async { return storage.GetTodos() }
      addTodo =
          fun todo ->
              async {
                  match storage.AddTodo todo with
                  | Ok () -> return todo
                  | Error e -> return failwith e
              } }

let webApp =
    Remoting.createApi ()
    |> Remoting.withRouteBuilder Route.builder
    |> Remoting.fromValue todosApi
    |> Remoting.buildHttpHandler

let app =
    application {
        url "http://0.0.0.0:8085"
        use_router webApp
        memory_cache
        use_static "public"
        use_gzip
    }

run app
```

with me pulling a pipe-operator syntax for C# from thin air, things could look something like this:

```csharp
var app = 
    args |> WebApplication.CreateBuilder
    builder |> AddRazorPages;
    builder |> Build;
    app |> app.Environment.IsDevelopment()
        ? UseDeveloperExceptionPage
        : UseExceptionHandler("/Error");
    app |> UseStaticFiles;
    app |> UseRouting;
    app |> UseAuthorization;
    app |> MapRazorPages;
    
app.Run();
```

There are lots of reasons that a pipe-forward operator would take a long time to get right, but if you think this would be helpful, I invite you do leave your thoughts on [this discussion](https://github.com/dotnet/csharplang/discussions/74).

**Comment by jmarolf** [comment]
As for @CyrusNajmabadi's actual proposal, I am curious what the LDM thinks. I don't think this is a bad addition, but I do not think there is overwhelming user pain today.

**Comment by CyrusNajmabadi** [comment]
> It sounds like what you want @najak3d (though it's always dangerous to assume here...) is a https://github.com/dotnet/csharplang/discussions/74.

Alternatively, i think that [auto-fluent for void methods](https://github.com/dotnet/csharplang/discussions/781) could work out for him.  That's something i'm *far* more likely to champion.

**Comment by jmarolf** [comment]
ah yes, that would be significantly less design work. and certainly would make things _better_.

**Comment by jmarolf** [comment]
I don't know what the scenario is where I will think that we _must_ have a forward-pipe operator. Maybe someday some important pattern (model-view-update?) will be so much better that we will be made to do it, but I am not seeing that happening in the short term.

**Comment by najak3d** [comment]
> > It sounds like what you want @najak3d (though it's always dangerous to assume here...) is a #74.
> 
> Alternatively, i think that [auto-fluent for void methods](https://github.com/dotnet/csharplang/discussions/781) could work out for him. That's something i'm _far_ more likely to champion.

IMO, "fluent-everywhere" is far more likely to cause mistakes/confusion than what I'm asking for.   In "fluent-everywhere" so you used to have a function that returned some *other* object, and so the code that uses it was expecting an entirely DIFFERENT object to be returned, and was operating on it...  Now you change the return type to "void" -- and there are NO COMPILER ERRORS -- the calling code now auto-operates upon the "void" which is interpreted to be the hosting object... Oops.   IMO, that is likely a very bad idea.

**Comment by najak3d** [comment]
> I don't know what the scenario is where I will think that we _must_ have a forward-pipe operator. Maybe someday some important pattern (model-view-update?) will be so much better that we will be made to do it, but I am not seeing that happening in the short term.

Thank you for your thoughtful response above.   Even for "rich" objects Serialization is heavily used in many settings.  Construction via serialization is everywhere.

What I'm wanting is simple, clear, and non-dangerous in nature.   It ONLY applies to the "Init Block" and simply makes the Init-Block work like a "Wither" block, with one caveat that the method calls must be at the end of the block, following the property setters.   So when compiler sees the first ".method()" statement, it completes construction, and the remainder of the methods calls are now operating on a constructed object.   It would be equivalent, exactly, to the long-hand notation.   It's simple, scoped short-hand, to aid in complex code-based compositions.

My request isn't truly "Fluent" -- it's a simple attractive replacement for Fluent, with notation that looks about the same, but does NOT require methods to return back the object instance to the caller.

This does NOT interfere ANY with Serialization concerns.  In fact, in some cases, it'll enhance Custom Serialization logic.  Imagine a custom serialization method like this:

```c#
MyType Deserialize(Reader reader)
{
   MyType obj = new MyType()
  {
            Name = reader.ReadString(),
            .ApplyStyleByName(reader.ReadString()),
            .SetPosition(reader.ReadInt32(), reader.ReadInt32()),
            .SetMargins(reader.ReadString())  // string representation of the Margins - this could be an extension method
  };
}
```

Serialization could simply choose whether or not they prefer this short-hand or not.   If not, then it makes zero impact on concerns for Serialization and Records, or PODs.

**Comment by najak3d** [comment]
Regarding serialization -- there is one mode of serialization that I've seen, which is high-performance.   Instead of saving data, it saves C# code, which is then compiled, and has the code that directly instantiates an object.

So just imagine a serializer that generates the C# that initializes each object, with the values directly inserted into the C#.   Then compile that, and load it as a dynamic DLL, and simply run it.

Backward compatibility becomes not too hard, because you approach it simply as you would approach backward compatibility for your API...   if your API is backward compatible, then the old serialized DLL's will also work for the new API. 

Other great thing about it is transparency and ease of debugging -- you aren't dealing with binary/string data... but you can simply look at it like raw code.   And even see compiler errors, where backward compatibility might be broken.

In ways, it's the best of many worlds.   I've used this approach for complex 3D scenes in the past.    Having nice C# syntax for construction would be useful for this type of serialization, especially.

**Comment by HaloFour** [comment]
@najak3d 

> So just imagine a serializer that generates the C# that initializes each object,

This would be a library concern, not a language concern.

**Comment by najak3d** [comment]
> @najak3d
> 
> > So just imagine a serializer that generates the C# that initializes each object,
> 
> This would be a library concern, not a language concern.

What that generated C# looks like is a language concern.   Is it succinct, or verbose?  Does it require extensions to make it work, or does C# syntax allow for the simpler syntax naturally?

**Comment by najak3d** [comment]
> > Instead of saving data, it saves C# code, which is then compiled, and has the code that directly instantiates an object
> 
> You have just re-invented [binary serialization](https://docs.microsoft.com/dotnet/standard/serialization/binary-serialization) from first principles. While this design is attractive in its simplicity of use it has several irreconcilable security flaws that have caused many real-world security exploits.
> 
> Security aside, what if you want to inter-operate with non-C# code? Being able to easily host a REST api that returns json is one of the cornerstones of why people use ASP.NET or microservices in general. The default design in C# cannot be "other forms of serialization are second class" when they represent 90% of what people are using serialization for.

Correct.  It has it's downsides.   For my experience, the Scene could also have Custom Code, already.  And so there truly was nothing you could accomplish in this sandboxed C# that you couldn't already accomplish via custom C#.   So for this scenario, that type of security wasn't an issue.

For MOST situations, you are 100% correct.   Which is surely a major reason why it's not used much.

In our case -- to achieve reasonable security, we just compiled the code run-time, so that we could do appropriate checks and confined the compiler environment to not-compile insidious code.  And so we serialized the Text, not the DLL directly.

That said -- this makes succinct syntax even more attractive for this scenario.

**Comment by HaloFour** [comment]
@najak3d 

> More succinct code, so long as it is just as clear - is preferable to more bloated syntax. 

If it's being compiled as a part of the project then it doesn't really matter, as long as it's clear.  The WPF designer doesn't have a problem here, and the language isn't going to adopt new syntax for the sake of making life slightly easier for designer code that is likely much easier to emit without trying to construct a complex fluent graph anyway.

**Comment by najak3d** [comment]
Like for Records, and concerns of immutability -- you could achieve this by making constructors that set all of the private fields, and then mark them "readonly".   Why do more?      Because doing more is helpful and beneficial by making it easier for us to have immutable classes.

**Comment by najak3d** [comment]
> @najak3d
> 
> > More succinct code, so long as it is just as clear - is preferable to more bloated syntax.
> 
> If it's being compiled as a part of the project then it doesn't really matter, as long as it's clear. The WPF designer doesn't have a problem here, and the language isn't going to adopt new syntax for the sake of making life slightly easier for designer code that is likely much easier to emit without trying to construct a complex fluent graph anyway.

I do agree that this is a minor concern.   Mostly because very little serialization is done in this fashion, and where it is being done, making the code shorter doesn't really help much.

I stepped into this because someone else suggested that "Fluent API might interfere with serialization concerns" -- and my response was simply "if anything, it can only help".   My proposal poses zero threat to his concerns over serialization.

The rest was just a brainstorm based on experience...  but you are right, that example is not a significant concern for us here.

**Comment by msedi** [comment]
Addtionally I would add some scenario that came up today:

I have a read-only structure where properties are dependent from other properties in the same structure. It would be completely sufficient to have everything set in the initialization using a compound initialization with init it is not possible for several reasons, because in the example order might play a role, so I cannot access the Blocksize within the initialization:

```cs
 var gridSize1 = new CudaGridBlockSize
            {
                BlocksizeX = 64,
                BlocksizeY = 1,
                BlocksizeZ = 1,
                GridsizeX = ((width - 1) / BlocksizeX / 4) + 1,
                GridsizeY = ((height - 1) / BlocksizeY) + 1,
                GridsizeZ = ((depth - 1) / BlocksizeZ) + 1
            } 
```

so I came up with the idea of using withers but found that also here I have no access to the original. Is this even possible?

```cs
 var gridSize1 = new CudaGridBlockSize
            {
                BlocksizeX = 64,
                BlocksizeY = 1,
                BlocksizeZ = 1
            } 
            with
            {
                // Accessing gridsize1 does not work
                GridsizeX = ((width - 1) / gridSize1.BlocksizeX / 4) + 1,
                GridsizeY = ((height - 1) / gridSize1.BlocksizeY) + 1,
                GridsizeZ = ((depth - 1) / gridSize1.BlocksizeZ) + 1
            };
```

The current solution would be:

```cs
 var gridSize1 = new CudaGridBlockSize
            {
                BlocksizeX = 64,
                BlocksizeY = 1,
                BlocksizeZ = 1
            } ;
  gridSize1 = gridSize1 with
            {
                GridsizeX = ((width - 1) / gridSize1.BlocksizeX / 4) + 1,
                GridsizeY = ((height - 1) / gridSize1.BlocksizeY) + 1,
                GridsizeZ = ((depth - 1) / gridSize1.BlocksizeZ) + 1
            };
```

which is not super complex but if you are working on this it might be something that also fits into this topic.

**Comment by totszwai** [comment]
Came here from an 2014 [SO](https://stackoverflow.com/questions/24951619/add-event-handler-during-object-initialization) thread LOL... 
Would be really nice to have such feature available, especially for those event ballbacks. 😭 

/subscribed

**Comment by acple** [comment]
This is exactly the feature I've been waiting for.
But how about Increment/Decrement operators? These are still useful in some common scenarios. Isn't it in scope?
```csharp
var countUp = myRecord with { Count++ };
```

**Comment by snarbies** [comment]
Just my two cents...

> could be interpreted as create `countUp` with the same count than `myRecord`, and then increment `myRecord.Count`.

I don't see why you wouldn't read this the same way then. 

```typescript
var countUp = myRecord with { Count += 1 };
```

They both convey exactly the same mutation on exactly the same storage location. I take it to be an issue of how we intuitively read it, but I think the fact that it's part of a `with` expression sufficiently clarifies.

**Comment by jnm2** [comment]
`++Count` means the same as `Count += 1`, but `Count++` has a slight difference:

```cs
var count = 0;
Console.WriteLine(count += 1); // 1
Console.WriteLine(++count); // 2
Console.WriteLine(count++); // Still 2
```

**Comment by snarbies** [comment]
They can mean something slightly different, yes, but not in this situation.

**Comment by snarbies** [comment]
Well... I can kind of see where it depends on how you look at it. But one interpretation makes sense, and one does not. We aren't keeping the value of the "expression", we're keeping the side-effect. 

In other words, if you interpret `with {count++}` to mean to modify the source, then how could you not interpret `with {count += 1}` the same way? They're both modifying the same thing.
